compiler/fir/fir2ir/src/org/jetbrains/kotlin/fir/backend/Fir2IrClassifierStorage.kt

```diff
@@ -89,6 +91,32 @@ class Fir2IrClassifierStorage(
                     ?: createIrTypeParameterWithoutBounds(original, index, context)
             }
         }
+        if (owner is FirCallableDeclaration<*>) {
+            val typeParameterInReturnType = owner.returnTypeRef.retrieveTypeParameter()
+            if (typeParameterInReturnType != null) {
+                getCachedIrTypeParameter(typeParameterInReturnType)
+                    ?: createIrTypeParameterWithoutBounds(typeParameterInReturnType, index = 0)
```

Comment:
This is strange place, as well as the next one. Normally we should not create type parameter when we don't know its index. This will anyway work incorrectly if a corresponding declaration has more than one type parameter.